open_spiel/games/gin_rummy.cc

```diff
@@ -0,0 +1,699 @@
+    if (action == kPassAction) {
+      deadwood_[cur_player_] = TotalCardValue(hands_[cur_player_]);
+      phase_ = Phase::kGameOver;
+    } else {
+      // Lay melds one action at a time.
+      SPIEL_CHECK_LE(action - kMeldActionBase, kNumMeldActions);
+      SPIEL_CHECK_GE(action - kMeldActionBase, 0);
+      layed_melds_[cur_player_].push_back(action - kMeldActionBase);
+      // Upon laying a meld the cards are removed from the player's hand.
+      std::vector<int> meld = int_to_meld.at(action - kMeldActionBase);
+      for (int card : meld)
+        RemoveFromHand(cur_player_, card);
+      deadwood_[cur_player_] = TotalCardValue(hands_[cur_player_]);
+      phase_ = Phase::kLayoff;
+    }
+  }
+}
+
+void GinRummyState::ApplyWallAction(Action action) {
+  if (action == kKnockAction) {
+    // When we've reached the wall, a knock automatically includes upcard.
+    UpcardToHand(cur_player_);
+    deadwood_[cur_player_] = MinDeadwood(hands_[cur_player_]);
+    // Make sure knock is legal.
+    SPIEL_CHECK_LE(deadwood_[cur_player_], knock_card_);
+    knocked_[cur_player_] = true;
+    prev_player_ = cur_player_;
+    phase_ = Phase::kKnock;
+  } else if (action == kPassAction) {
+    phase_ = Phase::kGameOver;
+  } else {
+    SpielFatalError("Invalid WallAction");
+  }
+}
+
+std::vector<Action> GinRummyState::LegalActions() const {
+  switch (phase_) {
+    case Phase::kDeal:
+      return DealLegalActions();
+    case Phase::kFirstUpcard:
+      return FirstUpcardLegalActions();
+    case Phase::kDraw:
+      return DrawLegalActions();
+    case Phase::kDiscard:
+      return DiscardLegalActions();
+    case Phase::kKnock:
+      return KnockLegalActions();
+    case Phase::kLayoff:
+      return LayoffLegalActions();
+    case Phase::kWall:
+      return WallLegalActions();
+    default:
+      return {};
+  }
+}
+
+std::vector<Action> GinRummyState::DealLegalActions() const {
+  std::vector<Action> legal_actions;
+  for (int card = 0; card < kNumCards; ++card) {
+    if (deck_[card] != kInvalidCard) legal_actions.push_back(card);
+  }
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::FirstUpcardLegalActions() const {
+  std::vector<Action> legal_actions;
+  // If both players have passed then must draw from stock.
+  if (pass_on_first_upcard_[0] && pass_on_first_upcard_[1]) {
+    legal_actions.push_back(kDrawStockAction);
+  } else {
+  // Options are to draw upcard or pass to opponent.
+  legal_actions.push_back(kDrawUpcardAction);
+  legal_actions.push_back(kPassAction);
+  }
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::DrawLegalActions() const {
+  std::vector<Action> legal_actions;
+  legal_actions.push_back(kDrawUpcardAction);
+  legal_actions.push_back(kDrawStockAction);
+  return legal_actions;
+}
+
+std::vector<Action> GinRummyState::DiscardLegalActions() const {
+  std::vector<Action> legal_actions;
+  for (int card : hands_[cur_player_]) {
+    legal_actions.push_back(card);
+  }
```

Comment:
isn't this just copy?

Reply:
Incorporated into new commit. This actually gave ~7% speedup in moves/s according to the benchmark_game test.